File: UI_testing/test_node_selenium/page_objects/base_element.js
const webDriver = require('selenium-webdriver')

class BaseElement {
    constructor(driver, _css, _xpath='') {
        this.driver = driver
        this._css = _css
        this._xpath = _xpath
    }

    async click() {
        await (await this.findElement()).click()
    }

    async goto(url) {
        await this.driver.get(url)
    }

    async screenshot() {
        await (await this.findElement()).takeScreenshot()
    }

    async findElement() {
        if (this._xpath == '') {
            return await this.driver.wait(webDriver.until.elementLocated(webDriver.By.css(this._css)))
        }
        else {
            return await this.driver.wait(webDriver.until.elementLocated(webDriver.By.xpath(this._xpath)))
        }
    }
}

module.exports = BaseElement

File: UI_testing/test_python_selenium/page_objects/base_element.py
from selenium.webdriver.support.ui import WebDriverWait

class BaseElement:
    def __init__(self, driver, browser, _css, _xpath):
        self.driver = driver
        self.browser = browser
        self.css = _css
        self.xpath = _xpath

    def click(self):
        self.find_element().click()

    def goto(self, url):
        self.browser.get(url)

    def screenshot(self, img_name):
        self.find_element()
        self.driver.save_screenshot(img_name)

    def find_element(self):
        if self.xpath == '':
            return WebDriverWait(self.driver).until(lambda x: x.find_element_by_css_selector(self.css))
        else:
            return WebDriverWait(self.driver).until(lambda x: x.find_element_by_css_selector(self.xpath))
